### PR TITLE
Allow --open option to specify the browser to use

### LINKED ISSUE #753: Specifying browser for --open flag
**Do you want to request a *feature* or report a *bug*?**

Feature

**What is the current behavior?**

Opens default browser which is not always dev browser

**What is the expected behavior?**

Adding ability to specify browser like

`webpack-dev-server --open chrome`

**If this is a feature request, what is motivation or use case for changing the behavior?**

Different environments

### PR DESCRIPTION
**Pull request by frankwinter**
Fixes #753

An additional parameter for the `--open` option will pass it to `opn`. For example `webpack-dev-server --open 'Google Chrome'` will open chrome browser. When the parameter is omitted the systems default browser starts up as usual.

### CODE REVIEW

File: bin/webpack-dev-server.js [hunk 1]
```diff
@@ -418,7 +418,12 @@ function reportReadiness(uri, options) {
 	if(options.historyApiFallback)
 		console.log(`404s will fallback to ${colorInfo(useColor, options.historyApiFallback.index || "/index.html")}`);
 	if(options.open) {
-		open(uri).catch(function() {
+		var params = [uri];
+
+		if(typeof options.open === "string")
+			params.push({ app: options.open });
+
+		open(...params).catch(function() {
```
Comment (SpaceK33z) [inline bin/webpack-dev-server.js]: Array spread doesn't seem to work in Node v4.

**Comment by sergeymorkovkin** [comment]
Any chances to have it merged? :)

**Comment by shellscape** [comment]
@sergeymorkovkin we need tests added to this PR for the new option. it'll become too difficult to manually test for each new version otherwise. @frankwinter needs to add that and resolve the conflicts recently introduces for this to get merged I'm afraid.

**Comment by sergeymorkovkin** [comment]
@shellscape Got it. Thanks for an update. If you think I could do this, please let me known.

File: .idea/inspectionProfiles/Project_Default.xml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<component name="InspectionProjectProfileManager">
```
Comment (michael-ciniawsky) [inline .idea/inspectionProfiles/Project_Default.xml]: Please blacklist and remove all the IDE specific files `.idea/*`

File: bin/webpack-dev-server.js [hunk 3]
```diff
@@ -333,8 +333,20 @@ function processOptions(wpOpt) {
...
 		options.compress = true;
 
-	if(argv["open"])
-		options.open = (argv["open"] === "") ? true : argv["open"];
+	if(argv["disable-host-check"])
+		options.disableHostCheck = true;
+
+	if(argv["open"] || argv["open-page"]) {
+		if(argv["open"] && argv["open"] !== "") {
+      options.open = argv["open"];
```
Comment (michael-ciniawsky) [inline bin/webpack-dev-server.js]: Indendation seems to be off here

**Comment by shellscape** [comment]
@frankwinter you'll have to remove the package-lock.json as well.

**Comment by frankwinter** [comment]
Sorry, please excuse the clumsy workflow.. It's my first contribution on a forked github-project :-/

File: bin/webpack-dev-server.js [hunk 4]
```diff
@@ -455,9 +459,15 @@ function reportReadiness(uri, options) {
 	if(options.historyApiFallback)
 		console.log(`404s will fallback to ${colorInfo(useColor, options.historyApiFallback.index || "/index.html")}`);
 	if(options.open) {
-		open(uri + options.openPage).catch(function() {
-			console.log("Unable to open browser. If you are running in a headless environment, please do not use the open flag.");
-		});
+		if(argv["open"] && argv["open"] !== "") {
+			open(uri + options.openPage, { app: options.open }).catch(function() {
```
Comment (shellscape) [inline bin/webpack-dev-server.js]: not a fan of two nearly identical calls to `open` and subsequent logging. should be something like this:

(this is pseudo code, don't copy and paste it)
```js
var openOptions = {};
var openMessage = "Unable to open browser";

if (typeof options.open === "string") {
  openOptions = { app: options.open };
  openMessage =+ " :  " + options.open;
}

open(uri + options.openPage, openOptions).catch(function () {
  console.log(openMessage + "If you are running in a headless environment, please do not use the open flag.");
});
```

File: bin/webpack-dev-server.js [hunk 5]
```diff
@@ -336,9 +336,13 @@ function processOptions(wpOpt) {
...
 		options.disableHostCheck = true;
 
-	if(argv["open"] || argv["open-page"]) {
+	if(argv["open-page"]) {
 		options.open = true;
-		options.openPage = argv["open-page"] || "";
+		options.openPage = argv["open-page"];
+	}
+
+	if(typeof argv["open"] !== "undefined") {
```
Comment (shellscape) [inline bin/webpack-dev-server.js]: why not

```js
if(argv["open"]) {
  options.open = argv["open"] || true;
}
```

**Comment by frankwinter** [comment]
If you call —open without a value argv["open“] will contain an empty string and options.open is going to be undefined.
Currently I don’t know if this can be managed somewhere else..

**Comment by shuijingleihen** [comment]
How to set the parameter in webpack.config.js file? My configure likes that:
```javascript
devServer: {
    port: '8081'
    ,compress: true
    ,watchContentBase: true
    ,open: 'chrome'
    ,openPage: 'views/index/'
}
```
but it doesn't work!!! And throw an error:

tim 20170808163019

**Comment by shellscape** [comment]
@shuijingleihen pull request discussions are not forums for support. Please ask questions on StackOverflow or the webpack Gitter (https://gitter.im/webpack/webpack).

**Comment by shuijingleihen** [comment]
@shellscape Oh sorry, I thought it was an issue, thank you for your reminding！

**Comment by tomchentw** [comment]
Any progress/changes on this PR? I'd love to help!

**Comment by shellscape** [comment]
@tomchentw I think we're good. @frankwinter needs to resolve the conflicts that popped up. I'll be on vaca until 8/27 but can merge when I'm back.

**Comment by tomchentw** [comment]
WOW!

**Comment by zigang93** [comment]
devServer: {
    port: '8081'
    ,compress: true
    ,open: 'firefox'
}

is working! thanks.  
Btw, should we update the doc on https://webpack.js.org/configuration/dev-server/#devserver-open ?

it should giving example of allow to use string and Boolean :)

**Comment by crowscript** [comment]
This solves the issue.
In package.json:
`"start": "webpack-dev-server  --config webpack.dev.js --open chrome",`

In webpack.config.js:
```
devServer: {
        ...
        open: 'chrome'
      },
```

**Comment by mrjdavidfg** [comment]
How can I open Firefox Developer Edition with the `open` property?

**Pull request merged**